src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneIndexingService.cs

```diff
@@ -0,0 +1,249 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using OrchardCore.ContentManagement;
+using OrchardCore.Entities;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Indexing;
+using OrchardCore.Lucene.Model;
+using OrchardCore.Modules;
+using OrchardCore.Settings;
+
+namespace OrchardCore.Lucene
+{
+    /// <summary>
+    /// This class provides services to update all the Lucene indices. It is non-rentrant so that calls 
+    /// from different components can be done simultaneously, e.g. from a background task, an event or a UI interaction.
+    /// It also indexes one content item at a time and provides the result to all indices.
+    /// </summary>
+    public class LuceneIndexingService
+    {
+        private const int BatchSize = 100;
+        private readonly IShellHost _shellHost;
+        private readonly ShellSettings _shellSettings;
+        private readonly LuceneIndexingState _indexingState;
+        private readonly LuceneIndexSettingsService _luceneIndexSettingsService;
+        private readonly LuceneIndexManager _indexManager;
+        private readonly IIndexingTaskManager _indexingTaskManager;
+        private readonly ISiteService _siteService;
+
+        public LuceneIndexingService(
+            IShellHost shellHost,
+            ShellSettings shellSettings,
+            LuceneIndexingState indexingState,
+            LuceneIndexSettingsService luceneIndexSettingsService,
+            LuceneIndexManager indexManager,
+            IIndexingTaskManager indexingTaskManager,
+            ISiteService siteService,
+            ILogger<LuceneIndexingService> logger)
+        {
+            _shellHost = shellHost;
+            _shellSettings = shellSettings;
+            _indexingState = indexingState;
+            _luceneIndexSettingsService = luceneIndexSettingsService;
+            _indexManager = indexManager;
+            _indexingTaskManager = indexingTaskManager;
+            _siteService = siteService;
+
+            Logger = logger;
+        }
+
+        public ILogger Logger { get; }
+
+        public async Task ProcessContentItemsAsync(string indexName = default)
+        {
+            // TODO: Lock over the filesystem in case two instances get a command to rebuild the index concurrently.
+            var allIndicesStatus = new Dictionary<string, int>();
+            var lastTaskId = Int32.MaxValue;
+            IDictionary indexSettingsList = null;
+
+            if (String.IsNullOrEmpty(indexName))
+            {
+                indexSettingsList = _luceneIndexSettingsService.List().Where(x => x.IndexInBackgroundTask).ToImmutableDictionary(x => x.IndexName, x => x);
```

Comment:
Here @jtkech. It's working like a charm.

Reply:
Yes, i assume because `LuceneIndexSettingsService` is a tenant singleton and then its `List()` returns a shared collection that might be mutated while iterating it.

So, here you would just need an isolated collection e.g by using a `ToList()`

    _luceneIndexSettingsService.List().ToList.Where( ... ).ToDictionary(x => x.IndexName, x => x);

But to not have to do the same elsewhere and because i think `ToList()` and `Where` (that use an enumerator) are not thread safe themselves, i think it is better to define in `LuceneIndexSettingsService` its **`_indexSettings` as an immutable array**. So that here you could just do

    _luceneIndexSettingsService.List().Where( ... ).ToDictionary(x => x.IndexName, x => x);

Then in `LuceneIndexSettingsService` you could do e.g

        public void CreateIndex(LuceneIndexSettings settings)
        {
            _indexSettings = _indexSettings.Add(settings);
            Update();
        }

Reply:
I think you are right because we could want to add an index while indexing an other one. Though I've had no issue with this scenario as I tested it. Every time we call the service it will create a new ImmutableDictionary from the ToList() call.

Reply:
I will try your suggestion as it would be cleaner.